lib/rules/no-unstable-nested-components.js

```diff
@@ -143,6 +143,31 @@ function isMapCall(node) {
   );
 }
 
+/**
+ * Check whether given node is part of a formatMessage call
+ * ```jsx
+ * {intl.formatMessage(messages.testMessage, {
+ * a: (...chunks) => <a href="/">{chunks}</a>
+ * })}
+ * ```
+ * @param {ASTNode} node The AST node
+ * @returns {Boolean} True if node is directly inside `map` call, false if not
+ */
+function isFormatMessageCall(node) {
+  return (
+    node
+    && node.parent
+    && node.parent.type === 'Property'
+    && node.parent.parent
+    && node.parent.parent.type === 'ObjectExpression'
+    && node.parent.parent.parent
+    && node.parent.parent.parent.type === 'CallExpression'
+    && node.parent.parent.parent.callee
+    && node.parent.parent.parent.callee.property
+    && node.parent.parent.parent.callee.property.name === 'formatMessage'
+  );
```

Comment:
this seems very hardcoded to one specific API from one specific hook. formatMessage could be renamed, and is `useIntl` a core hook, or a third-party one?

Reply:
It's a 3rd party library. I'm open to other suggestions on how to solve this, but I wanted to start with something that I know works. It's not a prop, so `allowAsProps` isn't relevant. I couldn't figure out how the `customValidators` work, but they seem to be related to `propTypes`, so that didn't seem to fit either.


I think having an option to just not traverse down into specific functions could cover both use cases. `ignoreFunctions: ['map', 'intl.formatMessage']`. What do you think?

Reply:
`intl.formatMessage` seems safe enough, but `map` is just *way* too broad.

imo the reality is that this is a very strange and unidiomatic API, and it simply conflicts with static analysis. I suspect that there could be changes in the useIntl library that would obviate this problem, but that's ofc not something either of us have control over.

Reply:
>  it simply conflicts with static analysis

It certainly conflicts with ESLint rules for static analysis. In my test code I included the call to `defineMessages` because I think it's good to have the whole context of how this is used. `defineMessages` doesn't do anything at runtime, it's exclusively there for static analysis/types to mark the places that need to be extracted for localization. It just returns the same object that's passed in. 

https://github.com/AsaAyers/eslint-plugin-react/blob/86e925a4e06fbdb8b37dbdbf42f0e10b5fb6b541/tests/lib/rules/no-unstable-nested-components.js#L231-L256